src/cudamatrix/cu-kernels.cu

```diff
@@ -741,6 +741,16 @@ static void _set_bias_params(Real* v, const Real* a, Real param_1, Real param_2,
   }
 }
 
+template<typename Real, typename OtherReal>
+__global__
+static void _copy_from_vec(int n, const Real* x, int incx, OtherReal* y, int incy) {
+  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x;
+   // if (blockIdx.y > 0) return;
+
+  if (i < n) {
+    y[1+(i-1)*incy] = (OtherReal) x[1+(i-1)*incx];
```

Comment:
The say, ", the performed operation is y [ j ] = α × x [ k ] + y [ j ] for
i = 1 , … , n , k = 1 + ( i - 1 ) *  incx and j = 1 + ( i - 1 ) *  incy ".

But this is when expressed in 1-based indexing, which is only an issue
because BLAS libraries are descended from FORTRAN so it's traditional to
talk that way.  The whole thing is more natural to express in zero-based
indexing, then it simplifies.
Dan




On Mon, May 23, 2016 at 2:23 PM, Ke Li <notifications@github.com> wrote:

> In src/cudamatrix/cu-kernels.cu
> <https://github.com/kaldi-asr/kaldi/pull/663#discussion_r64260078>:
>
> > @@ -741,6 +741,16 @@ static void _set_bias_params(Real* v, const Real* a, Real param_1, Real param_2,
> >    }
> >  }
> >
> > +template<typename Real, typename OtherReal>
> > +__global__
> > +static void _copy_from_vec(int n, const Real* x, int incx, OtherReal* y, int incy) {
> > +  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x;
> > +   // if (blockIdx.y > 0) return;
> > +
> > +  if (i < n) {
> > +    y[1+(i-1)*incy] = (OtherReal) x[1+(i-1)*incx];
>
> Sure, I'll change this soon. But I checked the cublas document and found
> that the i and j start with 1, for example:
> http://docs.nvidia.com/cuda/cublas/#cublas-lt-t-gt-copy. Is there a
> particular reason for this? Ke
> … <#m_2516107303449423798_>
> On Fri, May 20, 2016 at 2:29 PM, Daniel Povey ***@***.***> wrote: In
> src/cudamatrix/cu-kernels.cu <#663 (comment)
> <https://github.com/kaldi-asr/kaldi/pull/663#discussion_r64086749>>: > @@
> -741,6 +741,16 @@ static void _set_bias_params(Real* v, const Real* a, Real
> param_1, Real param_2, > } > } > > +template<typename Real, typename
> OtherReal> > +__global__ > +static void _copy_from_vec(int n, const Real*
> x, int incx, OtherReal* y, int incy) { > + int32_cuda i = blockIdx.x *
> blockDim.x + threadIdx.x; > + // if (blockIdx.y > 0) return; > + > + if (i
> < n) { > + y[1+(i-1)*incy] = (OtherReal) x[1+(i-1)*incx]; this looks like a
> bug: 1+(i-1)*incy should be i*incy. We always use zero-based indexing.
> [things like 1+(i-1)*incy might appear only where there was some weird
> conversion to/from one-based indexing.] — You are receiving this because
> you were mentioned. Reply to this email directly or view it on GitHub <
> https://github.com/kaldi-asr/kaldi/pull/663/files/86cab1989351bae78ec56730b008f7df8ec32834#r64086749>
>
> -- Ke Li Dept. of Electrical and Computer Engineering Johns Hopkins
> University Email: kli26@jhu.edu
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/663/files/86cab1989351bae78ec56730b008f7df8ec32834#r64260078>
>